tests/tests_pytorch/strategies/test_dp.py

```diff
@@ -123,17 +129,32 @@ def test_step(self, batch, batch_idx):
         self.add_outputs(output, batch.device)
         return output
 
-    def training_epoch_end(self, outputs):
-        assert outputs[0]["loss"].shape == torch.Size([])
-        self._assert_extra_outputs(outputs)
+    def training_step_end(self, training_step_output):
+        # the strategy does this automatically, but since we want to store these in memory, we need to manually do it
+        # so that we can append the reduced value and not the per-rank value
+        training_step_output["loss"] = self.trainer.strategy.reduce(training_step_output["loss"])
```

Comment:
Storing step-outputs in memory are not as simple with DP

If you do `self.train_outputs.append(output)` in the `training_step`, the outputs are per-rank and not reduced yet

So to replicate the old behaviour, one needs to manually reduce in `training_step_end` (which overrides the automatic reduction by the strategy: https://github.com/Lightning-AI/lightning/blob/fd61ed065a19b89f05b1d751ac1b9106cf68145c/src/lightning/pytorch/strategies/dp.py#L152-L153) so that you have a reference to append.

For evaluation, you still have to override `_step_end` but there's no need to reduce.